config/v1/types_infrastructure.go

```diff
@@ -500,6 +507,37 @@ type AWSResourceTag struct {
 	Value string `json:"value"`
 }
 
+// DNSConfigurationType contains information about API and API-Int Load Balancers
+// and who is responsible for that configuration.
+// +union
+type DNSConfigurationType struct {
+	// provider determines which DNS solution is in use for this cluster.
+	// When the user wants to use their own DNS solution, the `provider` is set to
+	// "UserProvided". In this state, the user is expected to pre-create the Load
+	// Balancers for API and API-Int.
+	// When the cluster is resposible for its own DNS configuration, this field is
+	// set to "ClusterProvided". The Load Balancers are also created by the cluster.
+	// The default is "ClusterProvided".
+	// +default="ClusterProvided"
+	// +kubebuilder:default:="ClusterProvided"
+	// +kubebuilder:validation:Enum="UserProvided";"ClusterProvided"
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="type is immutable once set"
+	// +unionDiscriminator
+	Provider DNSProviderType `json:"provider,omitempty"`
+}
+
+// DNSProviderType defines the source of the DNS and LB configuration.
+type DNSProviderType string
+
+const (
+	// DNSUserProvided indicates that the user provides the LB and DNS for API and API-Int.
+	DNSUserProvided DNSProviderType = "UserProvided"
+
+	// DNSClusterProvided indicates that the cluster provides the LB and DNS for API and API-Int.
+	DNSClusterProvided DNSProviderType = "ClusterProvided"
```

Comment:
These values seem a little squishy to me. I am having trouble keeping track of all the discussions around this feature, so I'm not sure if these values have been gamed out. 

My understanding of the intention here is that when the value is `UserProvided` that would indicate that we need to run `CoreDNS`. That to me has some cognitive dissonance. 

For the sake of discussion, what about something along the lines of
`CloudProvider` - this is standard IPI, DNS is in the cloud provider
`ClusterProvider` - uses on-cluster baremetal dns provider (coredns)
`UserProvider` - cluster/ipi does not provide DNS, this would be for something like ROSA where they are running dnsmasq in pods

Do these values better capture the current/future use cases we are aware of?

Reply:
I had a version that seperated 3 cases. But looking at it a different way, the cluster is taking care of DNS for both the CloudProvider and ClusterProvider case. I though I would get push back for superfluous cases and hence removed it.